Refactor k8s identities GC into a cell.Module

pippolo84 opened the pull request:
Port the current implementation of the CiliumIdentity GC to a separate `cell.Module`.

This makes possible to improve the testing of the GC code through unit tests, without relying on more complicated (and slow) controlplane tests.
Besides, the modularization of the GC cleans up the operator bootstrap code, since all the GC specific parts are now moved to their cell.

The new cell contains the code for both GC modes: CRD allocated identities and kvstore allocated identities.
The GC is moved inside the `identity` package and has the following structure:

```
operator/identity/
├── cell.go
├── crd_gc.go
├── gc.go
├── heartbeat.go
└── kvstore_gc.go
```

where:

- `cell.go` contains all the exported API for the new cell, alongside its dependencies and configuration.
- `gc.go` contains the common part of both gc modes.
- `crd_gc.go` and `heartbeat.go` contain the CRD allocation mode specific GC part.
- `kvstore`contains the kvstore allocation mode specific GC part.

The new cell relies on `resource.Resource[*v2.CiliumIdentity]` to listen for CiliumIdentity related events and react accordingly.

**Note for the reviewers**

This is probably better to review commit by commit. The most impactful one is 1a482f8d3a08147410e65d91adbc6e49515a49b4, but despite the number of files modified, it aims to add the minimum amount of changes needed to modularize the identities GC code in a separate cell.
Further cleanups are left in the subsequent commits.

pippolo84 committed:
k8s: Add shared resource for Cilium Identities

Add Cilium Identities as a shared resource in the proper cell.  This
will enable the modularization of all subsystems that depend on
CiliumIdentities events streaming.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on operator/cmd/k8s_identity_gc_test.go:
```diff
@@ -0,0 +1,199 @@
+			func(exitCode int) {
+				// Force garbage-collection to force finalizers to run and catch
+				// missing Event.Done() calls.
+				runtime.GC()
```
I don't think we really need this outside `pkg/k8s/resource`

joamaki commented on operator/cmd/k8s_identity_gc_test.go:
```diff
@@ -0,0 +1,199 @@
+
+		// To ignore leaked goroutine started from the instantiation
+		// of global variable watchers.nodeQueue
+		goleak.IgnoreTopFunction("k8s.io/client-go/util/workqueue.(*delayingType).waitingLoop"),
```
We shouldn't ignore this. It's due to not having called `workqueue.ShutDown` (or was it `ShutdownWithDrain`...). Not sure where it's from though.

pippolo84 replied:
It's from [here](https://github.com/cilium/cilium/blob/48ddc9a494c756a7f6ba03b93bd1715dbd113f99/operator/watchers/node.go#L34).
Being an unexported package-level variable it is not possible to call `ShutDown`/`ShutDownWithDrain` from the test.
Exported APIs that close that queue are only:

- `RunCiliumNodeGC`
- `HandleNodeTolerationAndTaints`

But I think calling either of these is gonna pollute the test code even more than ignoring the leak.
Best I can do is to add an exported API (to be called from unit tests) to the watchers package.
Once the watchers package will be refactored and cleaned from global variables, we will get rid of that. WDYT?

joamaki replied:
Ah I see. Yep sounds good!

joamaki commented on operator/cmd/k8s_identity_gc_test.go:
```diff
@@ -0,0 +1,199 @@
+
+	hive := hive.New(
+		// provide a mock clientset
+		cell.Provide(func() k8sClient.Clientset {
```
You can also flip this around and use `resource.FakeClientSet` and then have an `Invoke` that pulls it out, or sets up the initial state.

joamaki commented on operator/cmd/k8s_identity_gc_test.go:
```diff
@@ -0,0 +1,199 @@
+		cell.Provide(newTestIdentitiesGCConfig),
+
+		// force instantiation of identity-gc to start the test
+		cell.Invoke(func(*IdentitiesGC) {}),
```
For "standalone" things we probably should just use an `Invoke` that registers the start and stop hooks and not provide an object that nobody uses. Only drawback is that it's slightly harder to visualize it (e.g. dot-graph), but we can solve that separately.

joamaki commented on operator/cmd/k8s_identity.go:
```diff
@@ -10,78 +10,208 @@ import (
-	} else {
-		log.WithField(logfields.Identity, identity.GetName()).Debug("Updated identity")
+// IdentitiesGC represents the Cilium identities periodic GC.
+type IdentitiesGC struct {
```
This can be private, e.g. `identitiesGC`

joamaki commented on operator/cmd/k8s_identity.go:
```diff
@@ -10,78 +10,208 @@ import (
+
+// IdentityGCDeps contains all the dependencies for the identity-gc.
+// They will be provided through dependency injection.
+type IdentityGCDeps struct {
```
This can be private, `identityGCParams`. Use `*Params` as elsewhere.

joamaki commented on operator/cmd/k8s_identity.go:
```diff
@@ -10,78 +10,208 @@ import (
+	cell.In
+
+	Logger logrus.FieldLogger
+	LC     hive.Lifecycle
```
Use `Lifecycle` instead since none of the other field names are shortened.

joamaki commented on operator/cmd/k8s_identity.go:
```diff
@@ -10,78 +10,208 @@ import (
-	err := identityRateLimiter.Wait(ctx)
-	if err != nil {
-		return err
+	cell.Provide(newIdentitiesGCConfig),
```
All `IdentityGC*` params shoud be moved into `IdentityGCConfig` from operator config, and we should then do `cell.Config(defaultIdentityGCConfig)` here. We'll still have few options that will be from operator config, but eventually they should also be split into per-module configs.

I guess the thing stopping this is that we need `IdentityGCInterval` also in the kvstore identity GC? It would probably share the same configuration?

pippolo84 replied:
Exactly. But I managed to move the kvstore mode gc code into this cell, so that option can be removed from the global operator config.

joamaki commented on operator/cmd/k8s_identity.go:
```diff
@@ -10,78 +10,208 @@ import (
-	if err != nil {
-		return err
+	cell.Provide(newIdentitiesGCConfig),
+	cell.Provide(newIdentitiesGC),
```
Consider instead `cell.Invoke(registerIdentitiesGC)` which just adds the lifecycle hooks. We don't really need the `IdentitiesGC` object in the object graph.

pippolo84 replied:
I'll go for an unexported struct, in order to encase all the required data for the gc to operate (logger, clientset, interval, rate limiting, etc.)

joamaki commented on operator/cmd/k8s_identity.go:
```diff
@@ -10,78 +10,208 @@ import (
+	igc.cancel = cancel
+
+	// start heartbeat updater
+	igc.wg.Add(1)
```
Consider using `cilium/workerpool` instead.

joamaki commented on operator/cmd/k8s_identity.go:
```diff
@@ -10,78 +10,208 @@ import (
+	events := igc.identityResource.Events(ctx)
+	for {
+		select {
+		case <-ctx.Done():
```
No need for this case. The events channel closes when ctx is cancelled, e.g. do:
`for event := range events {` instead.

joamaki commented on operator/cmd/k8s_identity.go:
```diff
@@ -10,78 +10,208 @@ import (
+
+	cfg IdentitiesGCConfig
+
+	clientset        k8sClient.Clientset
```
using the subclient here might be cleaner? e.g.  interface returned by `clientset.CiliumV2().CliiumIdentities()`.

joamaki commented on operator/cmd/k8s_identity_gc_test.go:
```diff
@@ -0,0 +1,199 @@
+	// signal that endpoints are sync-ed, otherwise identities gc won't start
+	close(watchers.CiliumEndpointsSynced)
+
+	if err := retryN(
```
I'd probably prefer just spelling out the retry loop here: `for retry := 0; retry < 10; retry++ { list(); ...; time.Sleep(...) }`

Side note: We should not use the real clock but inject a fake clock that we control. Otherwise our tests will be way too slow. For this it'd mean that `controllers` would need to use that. If we'd expose `pkg/controllers` as a module it could depend on a global clock that our tests mock.

joamaki commented on operator/cmd/k8s_identity.go:
```diff
@@ -10,78 +10,208 @@ import (
+
+func newIdentitiesGCConfig(daemonCfg *option.DaemonConfig, operatorCfg *operatorOption.OperatorConfig) IdentitiesGCConfig {
+	return IdentitiesGCConfig{
+		AllocationMode:           daemonCfg.IdentityAllocationMode,
```
What is `DaemonConfig` doing here? Do we have lots of places in the operator that pull that in? Worth perhaps making a list and fixing them soon.

pippolo84 replied:
Solved using a separate structure to avoid pollution with global config options. I'm gonna make a list of such config options "cross-references".

joamaki commented on operator/cmd/k8s_identity.go:
```diff
@@ -10,78 +10,208 @@ import (
+	}
+
+	if igc.cfg.EndpointGCInterval == 0 {
+		log.Fatal("The CiliumIdentity garbage collector requires the CiliumEndpoint garbage collector to be enabled")
```
Hmm, any idea why this is necessary? I wonder if things like this (e.g. cross-module dependencies) should be solved outside, e.g. in `root.go`.

pippolo84 replied:
Moved the check outside, in `operator/cmd/root.go`. TBH, I don't know why this is necessary, though. Needs to investigate.

joamaki commented on operator/cmd/k8s_identity.go:
```diff
@@ -10,78 +10,208 @@ import (
+				RunInterval: igc.cfg.IdentityGCInterval,
+				DoFunc: func(ctx context.Context) error {
+					igc.gc(ctx)
+					return ctx.Err()
```
Why do we return `ctx.Err()`? Don't quite see how that's useful.

pippolo84 replied:
Just got it from [previous implementation](https://github.com/cilium/cilium/blob/274614fa66b43cf5898fba1164a64b16eafb93b3/operator/cmd/k8s_identity.go#L188), but I agree it's not useful. I'll remove it.

joamaki commented on operator/cmd/k8s_identity.go:
```diff
@@ -91,157 +221,117 @@ func identityGCIteration(ctx context.Context, clientset k8sClient.Clientset) {
-
+	for _, identity := range identities.Items {
 		// The identity is definitely alive if there's a CE using it.
 		if watchers.HasCEWithIdentity(identity.Name) {
```
Unfortunate that we're still depending on this. Perhaps worth as next step to bring in `CiliumEndpoints` as shared resource.

joamaki commented on operator/cmd/k8s_identity_gc_test.go:
```diff
@@ -0,0 +1,199 @@
+	}
+
+	// signal that endpoints are sync-ed, otherwise identities gc won't start
+	close(watchers.CiliumEndpointsSynced)
```
Should add `Resource[*CiliumEndpoints]` and get rid of this. The test also doesn't seem to check whether we correctly keep identities that have a CiliumEndpoint associated with them.

joamaki replied:
We'll also need to have the CiliumEndpoints indexed by identity ID (e.g. with something like: https://github.com/joamaki/cilium/blob/pr/joamaki/operator-resource-trials/pkg/k8s/resource/index.go).

pippolo84 replied:
Yep, I'll make this the priority for the next architecture-related PR.

joamaki commented on operator/identity/gc_cell.go:
```diff
@@ -0,0 +1,223 @@
+// The GC subscribes to identities events to mark all the related identities
+// as alive. If an identity has no activity for a prolonged interval,
+// it is first marked for deletion and eventually deleted.
+var IdentitiesGCCell = cell.Module(
```
I would suggest putting hive/cell related definitions into `cell.go` and the actual implementation code into crd_gc.go (e.g. the `identityGC` struct and so on). Idea would be to have `cell.go` just contain the exported cell, config struct and the API interface type so it's quick to scan what this thing needs and what it's providing.

joamaki commented on operator/identity/gc_cell_test.go:
```diff
@@ -0,0 +1,199 @@
+		}),
+
+		// provide a identities resource events
+		cell.Provide(func(lc hive.Lifecycle, c k8sClient.Clientset) resource.Resource[*v2.CiliumIdentity] {
```
You can also just directly use `k8s.SharedResourcesCell` here. `Resource` is lazy and won't do anything unless `Events()` or `Store()` is called.

pippolo84 committed:
test/controlplane: Add support for cell specific configs

Unlike global configuration options, cell-specific configuration options
(i.e. the ones defined through cell.Config(...)) will not be loaded from
`agentOption` or `operatorOption`, but from the `*viper.Viper` object
bound to the test hive.

`modCellConfig` function exposes the operator hive viper struct to each
controlplane test, so to allow changing those options as needed.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on operator/identity/cell.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+func (def IdentitiesGCConfig) Flags(flags *pflag.FlagSet) {
+	flags.Duration(operatorOption.IdentityGCInterval, def.IdentityGCInterval, "GC interval for security identities")
```
Hmm, if it's easy it might be worth moving `operatorOption.IdentityGCInterval` etc. definitions here. I think it'll end up being confusing soon when more things move out from OperatorConfig but the flag names remain there.

joamaki commented on operator/identity/gc_test.go:
```diff
@@ -0,0 +1,177 @@
+
+		// initial setup for the test
+		cell.Invoke(func(clientset k8sClient.Clientset) {
+			nodes := []*corev1.Node{
```
This is big enough that it warrants to be a named top-level function.

joamaki commented on operator/identity/gc_test.go:
```diff
@@ -0,0 +1,177 @@
+			}
+		}),
+
+		cell.Invoke(func(lc hive.Lifecycle, clientset k8sClient.Clientset, gc *identitiesGC) {
```
`clientset` and `gc` don't seem to be referenced here. Could also move this code into the test setup function.

pippolo84 committed:
operator/identity: Do not export identities heartbeat API

The heartbeat store used to keep track of the last insert/update
timestamps of each identities is an implementation detail related only
to the inner workings of the identities GC.

The commit changes all the related APIs to make them unexported.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator: Refactor cilium identities GC to a cell

Add a new cell.Module to implement the CiliumIdentity GC.

This improves the testability of the GC code through unit tests, without
relying on more complicated (and slow) controlplane tests.
Besides, the modularization of the GC cleans up the operator bootstrap
code, since all the GC specific parts are now moved to their cell.

The cell supports both GC modes: CRD allocated identities or kvstore
allocated identities.
The code is moved inside the `identity` package and has the following
structure:

operator/identity/
├── cell.go
├── crd_gc.go
├── gc.go
├── heartbeat.go
└── kvstore_gc.go

where:

- `cell.go` contains all the exported API for the new cell, alongside
  its dependencies and configuration.
- `gc.go` contains the common part of both gc modes.
- `crd_gc.go` and `heartbeat.go` contain the CRD allocation mode
  specific GC part.
- `kvstore`contains the kvstore allocation mode specific GC part.

The new cell relies on `resource.Resource[*v2.CiliumIdentity]` to listen
for CiliumIdentity related events and react accordingly.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator/identity: Add a unit test for GC

Add a unit test for the CiliumIdentities GC operating in CRD allocation
mode.

The new unit test verifies that an Identity is correctly gc-ed when it
has neither been updated in the last heartbeat cycle nor used by a
CiliumEndpoint.
Besides, the test adds a check for leaked goroutines using goleak.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator: Export API to clean up nodequeue goroutine

The global node watcher relies on an unexported rate limiting queue that
starts a goroutine to listen for related events.
Being unexported, there are no means to stop that goroutine when running
unit tests like the identities gc one in operator/identity.
To avoid leaking goroutines, an exported API is added so to stop the
node queue before the test termination.

This is a temporary solution before refactoring the watchers package to
use the new "hive and cells" architecture.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
test/controlplane: Remove identities GC test

Controlplane test identities GC has been superseded by the identities GC
unit test in operator/identity. The two test are functionally
equivalent, so the controlplane one is removed.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on pkg/k8s/shared_resources.go:
```diff
@@ -9,6 +9,7 @@ import (
 
 	"github.com/cilium/cilium/pkg/hive"
 	"github.com/cilium/cilium/pkg/hive/cell"
+	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
```
nit: use `cilium_v2`

joamaki commented on operator/identity/crd_gc.go:
```diff
@@ -0,0 +1,184 @@
+			})
+	}
+
+	return igc.wp.Submit("hearbeat-updater", igc.runHeartbeatUpdater)
```
s/hearbeat/heartbeat/

Shouldn't this be only started when `gcInterval != 0`?

pippolo84 replied:
I thought the same but avoided to do so because the previous implementation was starting the heartbeat-updater even when `gcInterval == 0`. Anyway, it seems to be something that matters only to the identities gc inner workings, so I guess it was unintended.
I'll add a separate commit to highlight the change from previous implementation.

joamaki commented on operator/identity/crd_gc.go:
```diff
@@ -0,0 +1,184 @@
+
+	identities, err := igc.clientset.List(ctx, metav1.ListOptions{})
+	if err != nil {
+		igc.logger.WithError(err).Error("unable to list Cilium identities")
```
This method should always return any errors it encounters in order to mark the controller as failing.

joamaki commented on operator/identity/crd_gc.go:
```diff
@@ -0,0 +1,184 @@
+		return nil
+	}
+
+	identities, err := igc.clientset.List(ctx, metav1.ListOptions{})
```
Don't use `clientset.List` since we already have the identity resource which maintains a local store. Use `igc.identity.Store()`.

joamaki replied:
Or extend IdentityHeartbeatStore to contain all required state for garbage collection?

pippolo84 replied:
Probably using `igc.identity.Store()` is a bit simpler, since we also have to check that each identity is not used by an Endpoint.

joamaki commented on operator/identity/crd_gc.go:
```diff
@@ -0,0 +1,184 @@
+			}).Debugf("Deleting unused identity; marked for deletion at %s", ts)
+
+			if err := igc.deleteIdentity(ctx, &identity); err != nil {
+				log.WithError(err).WithFields(logrus.Fields{
```
Since pkg/controller is invoking this method, it'd be good to collect all errors and return them so we can see the controller as failing. Same also with updateIdentity() on line 98.

joamaki commented on operator/identity/crd_gc.go:
```diff
@@ -0,0 +1,184 @@
+			},
+		},
+	); err != nil {
+		log.WithError(err).Error("Unable to delete identity")
```
No need to log errors here as we log it at call site.

joamaki commented on operator/identity/crd_gc.go:
```diff
@@ -0,0 +1,184 @@
+		identity,
+		metav1.UpdateOptions{},
+	); err != nil {
+		log.WithError(err).Error("Updating Identity")
```
No need to log here.

joamaki commented on operator/identity/crd_gc.go:
```diff
@@ -0,0 +1,184 @@
+		return err
+	}
+
+	log.WithField(logfields.Identity, identity.GetName()).Info("Garbage collected identity")
```
Do we need to log this? Seems potentially quite spammy.

pippolo84 replied:
It surely could be. Probably the right trade-off is to lower it to `Debug` level.

joamaki commented on operator/identity/gc.go:
```diff
@@ -0,0 +1,157 @@
+	case option.IdentityAllocationModeKVstore:
+		return igc.startKVStoreModeGC(ctx)
+	default:
+		igc.logger.Fatalf("Unknown Cilium identity allocation mode: %q", igc.allocationMode)
```
Don't fatal in Start(), return the error instead to gracefully abort start.

joamaki commented on operator/identity/gc.go:
```diff
@@ -0,0 +1,157 @@
+
+func newGC(
+	params gcParams,
+	cfg GCConfig,
```
cfg and sharedCfg should be in `gcParams`

joamaki commented on operator/identity/kvstore_gc.go:
```diff
@@ -0,0 +1,97 @@
+	log.WithField(logfields.Interval, igc.gcInterval).Info("Starting kvstore identity garbage collector")
+	backend, err := kvstoreallocator.NewKVStoreBackend(cache.IdentitiesPath, "", nil, kvstore.Client())
+	if err != nil {
+		igc.logger.WithError(err).Fatal("Unable to initialize kvstore backend for identity allocation")
```
Don't fatal, return the error.

joamaki commented on operator/identity/gc_test.go:
```diff
@@ -0,0 +1,298 @@
+
+		// initial setup for the test
+		cell.Invoke(func(c k8sClient.Clientset) {
+			clientset = c
```
Move into the invoke below?

joamaki commented on operator/identity/cell.go:
```diff
@@ -0,0 +1,96 @@
+// The GC subscribes to identities events to mark all the related identities
+// as alive. If an identity has no activity for a prolonged interval,
+// it is first marked for deletion and eventually deleted.
+var GCCell = cell.Module(
```
Are there other cells? Can this be just `Cell`? Same for everything else with `GC` prefix, though maybe the package should be renamed to identitygc.

joamaki commented on operator/identity/cell.go:
```diff
@@ -0,0 +1,96 @@
+	cell.Provide(newGC),
+
+	// Invoke forces the instantiation of the identity gc
+	cell.Invoke(func(*GC) {}),
```
Why not just `cell.Invoke(newGC)`? I don't see any exported methods in `GC` struct. Though perhaps preferably it'd just be `registerIdentityGC` and it'd just append the lifecycle hooks and not return an internal object.

joamaki commented on operator/identity/gc_test.go:
```diff
@@ -0,0 +1,298 @@
+	sharedCfg GCSharedConfig,
+) {
+	// create the gc (it will be managed through hive lifecycle hooks)
+	_ = newGC(params, cfg, sharedCfg)
```
I'd not bury this in here. You can just do `cell.Invoke(newGC)` inside `hive.New` above. And rename this method to `setupCiliumEndpointWatcher` or something.

pippolo84 committed:
operator/identity: Disable heartbeat when interval is 0

Do not start the heartbeat updater for CRD mode identities GC when the
related interval is set to 0.

The heartbeat updater single purpose is to mark all identities that have
been just created, in order to avoid the GC to collect them immediately
after creation.  When the GC is disabled, there is no point in starting
it.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator/identity: Fix logging of CRD mode GC

To avoid repeating log entries, inner logging of updates and deletions
ops on identities is removed.

Also, the log entry related to the collection of an identity is lowered
to debug level to avoid flooding the log too much.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator/identity: Avoid Fatal() while starting GC

In order to gracefully shutdown the identities GC, all critical errors
are reported to the upper layer instead of relying on a call to Fatal().

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator/identity: Rename package to identitygc

All the code in operator/identity is there only to implement the Cilium
identities gc. Renaming it makes its purpose clearer and also avoids the
name clashing with pkg/identity package.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator/identitygc: Use a register function to invoke gc

Since the GC does not export any API, it is possible not to return the
internal object and just register its start and stop hooks to
hive.Lifecycle directly.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

tommyp1ckles commented on operator/identitygc/crd_gc.go:
```diff
@@ -0,0 +1,189 @@
+	select {
+	case <-watchers.CiliumEndpointsSynced:
+	case <-ctx.Done():
+		return nil
```
should return `ctx.Err()`?

pippolo84 replied:
The old code was doing that, but we were discussing [here](https://github.com/cilium/cilium/pull/22892#discussion_r1064469239) with Jussi that that's not much useful (basically it will always return `context.Canceled`).

tommyp1ckles commented on operator/identitygc/crd_gc.go:
```diff
@@ -0,0 +1,189 @@
+					logfields.K8sUID:   identity.UID,
+				}).Info("Marking identity for later deletion")
+
+				if identity.Annotations == nil {
```
If `identity.Annotations == nil` this will have already crashed at line 91, we should just ensure its not null/construct this there.

pippolo84 replied:
Actually, it's safe to check for the existence of a key in a nil map. Only a write is gonna panic.
If the key does not exist or the map is nil, we will end up in this branch and in the latter case we'll lazily create the map.

tommyp1ckles commented on operator/identitygc/gc.go:
```diff
@@ -0,0 +1,148 @@
+
+	enableMetrics bool
+	// counters for GC failed/successful runs
+	failedRuns     float64
```
these should probably just be ints

pippolo84 replied:
I've used floats because that's what `(prometheus.Gauge).Set` requires, but you're right, this may feel confusing. Better to change this and cast that at `Set` call site.

pippolo84 committed:
identitygc: Use int types for prometheus metrics

Even if (prometheus.Gauge).Set requires a float64 value, declaring the
counters for successful and unsuccessful gc runs as such might feel
confusing to the casual reader.  The type is then changed to int and a
type conversion to float64 is added at each Set call site.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
identitgygc: Deep copy the identities before updating them

The identities returned by the underlying identity resource informer are
owned by the informer itself, thus they should be deep copied before
being modified.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
docs: Regenerate cmdref following identities gc refactor

This commit regenerates the cmdref documentation files following the
refactor of the identites gc into a cell.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>
